be/src/util/compression/compression_context_pool.h

```diff
@@ -0,0 +1,128 @@
+// This file is made available under Elastic License 2.0.
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include <shared_mutex>
+
+#include "common/status.h"
+#include "common/statusor.h"
+
+namespace starrocks::compression {
+
+/**
+ * A compression context pool preserves many unused compression context.
+ *
+ * The caller can apply a compression context from the pool through `get()` and then use it.
+ * When the compression context is finished being used, it will automatically call the delete
+ * function of unique_ptr `ReturnToPoolDeleter` and then return the context to the pool.
+ *
+ * Under normal circumstances, compression context will not be cleaned up, if you want to free
+ * all the context in the pool, you can call `flush_deep()`.
+ */
+template <typename T, typename Creator, typename Deleter, typename Resetter>
+class CompressionContextPool {
+private:
+    using InternalRef = std::unique_ptr<T, Deleter>;
+
+    class ReturnToPoolDeleter {
+    public:
+        using Pool = CompressionContextPool<T, Creator, Deleter, Resetter>;
+
+        explicit ReturnToPoolDeleter(Pool* pool) : _pool(pool) { DCHECK(pool); }
+
+        void operator()(T* t) {
+            InternalRef ptr(t, _pool->_deleter);
+            _pool->add(std::move(ptr));
+        }
+
+    private:
+        Pool* _pool;
+    };
+
+public:
+    using Object = T;
+    using Ref = std::unique_ptr<T, ReturnToPoolDeleter>;
+
+    explicit CompressionContextPool(Creator creator = Creator(), Deleter deleter = Deleter(),
+                                    Resetter resetter = Resetter())
+            : _creator(std::move(creator)),
+              _deleter(std::move(deleter)),
+              _resetter(std::move(resetter)),
+              _stack(),
+              _created(0) {}
+
+    StatusOr<Ref> get() {
+        std::unique_lock wrlock(_stack_lock);
+        if (_stack.empty()) {
+            StatusOr<T*> t = _creator();
```

Comment:
`ASSIGN_OR_RETURN` ?

Reply:
Maybe you could call the `_creator` without a lock, which may take a lot of time?